### DISCUSSION TITLE
Where is Langium placed compared to XText,Xtend,Maven,Gradle

### DISCUSSION THREAD

**Discussion by mehmetsalgar** [eclipse-langium/langium#144] [Q&A]
Hi everybody,

recently I was researching about the feature of the XText and potential frameworks to replace it and I found the web page for Langium but I can't place it which role it is exactly playing.

At first it seems, it is replacing XText but the biggest attraction of XText for me, it is deep integration with XTend for code generation but I can't see any component in Langium for code generation (may be it is naturally matched with TypeScript but I am not sure, my knowledge with TypeScript is rudimentary).

I also see no integration with Maven/Gradle, which is quite important for my use cases (we are generating code from our CI pipelines and not from Eclipse UI).

And finally, my main motivation for looking what happening after XText, the XText's close integration with Eclipse (My project uses exclusively IntelliJ and it is reducing the acceptance of using DSL for scenarios that will make sense) and .p2 repositories are pain in the ass for CI Integration over Maven/Gradle....

Now what I am going to write is pure speculation but is the purpose of Langium is to provide a Language Server from DSL Grammar so we can connect with, lets say, IntelliJ and work from there?

Thx for answers?

**Comment by msujew** [comment]
Hi @mehmetsalgar,

Regarding code generation, we are relying on a mixture of [TypeScripts template strings](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) and our own generator utility, which we are currently in the process of improving (see #133). We don't plan to integrate another third party language into that workflow.

Maven/Gradle are tools mainly for the JVM, while we work exclusively with the node.js/web stack. For that reason, we provide everything via `npm`, as you can already see with our various packages ([langium-core](https://www.npmjs.com/package/langium), [langium-cli](https://www.npmjs.com/package/langium-cli), [yeoman generator](https://www.npmjs.com/package/generator-langium)). We recommend users to implement their own (code)-generators via a CLI, which is able to run anywhere, independent of any IDE or runtime environment (only depending on node.js).

You should be able to set up IntelliJ with a language server that is provided by Langium. We don't plan to support that use case out-of-the-box, but it should be possible. You would probably have to create a script file (bash or similiar) to actually start the language server. In our examples, vscode takes care of that. Xtext actually does something similiar with its own language servers.

Your speculation is correct. Langium is built with the language server protocol in mind. Compared to Xtext, we omit most abstraction layers between our own framework and the lsp and work directly on the lsp structure internally.

Does that answer all of your questions?

**Reply by mehmetsalgar** [reply]
To say the truth, I am little bit lost at the moment, I guess in time, you will have more examples, so I might get a better grip.

From your 'Hello World' example, I can't find my way how to generate code or integrate with CI....

**Reply by msujew** [reply]
Yeah, that's a bit of an issue right now. We have an open issue that will be resolved in the coming weeks #131 for exactly that type of problem.

This should also be a part of the general Langium documentation, which we weren't able to work on yet. That's also coming in the next few weeks.

**Reply by msujew** [reply]
For now, you can take a look at how the [`langium-cli`](https://github.com/langium/langium/tree/main/packages/langium-cli) package works, to get an idea, what I'm talking about.

**Reply by mehmetsalgar** [reply]
So this is where my Template should be defined for code generation?

`
"textMate": {
            "out": "syntaxes/domain-model.tmLanguage.json"
        }`

**Reply by mehmetsalgar** [reply]
On second tought, I think not, I am trying to find too much similiraties between MWE Workflows and langium I think....

**Reply by msujew** [reply]
No, that's just our own metadata to tell the `langium-cli` package which path the generated textMate syntax should have.

Take a look at the following code to see how the code generation utilities from Langium are meant to be used within a CLI package:

https://github.com/langium/langium/blob/38ffd3ea6c00aa7bfc525adc8223b236cb50916f/packages/langium-cli/src/generator/ast-generator.ts#L70-L101

It's not as simple as Xtend, but that could still be improved.
